extern/cryptoki.h

```diff
@@ -0,0 +1,65 @@
+#ifndef _CRYPTOKI_H_
+#define _CRYPTOKI_H_ 1
+
+/* MIT License */
+
+/* Copyright (c) 2019 Eric Devolder */
+
+/* Permission is hereby granted, free of charge, to any person obtaining  */
+/* a copy of this software and associated documentation files (the        */
+/* "Software"), to deal in the Software without restriction, including    */
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND                  */
+/* NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE */
+/* LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION */
+/* OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION  */
+/* WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.        */
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#if defined(__CYGWIN64__)
+#pragma warning "Cygwin 64 bits build will only work with Cygwin64-compiled PKCS#11 modules"
+#endif
+
+#define CK_PTR            *
```

Comment:
So I understood these defines were different on the different platforms. Is that not the case?

Reply:
Short answer: not really...

Long one - take a deep breath:
PKCS#11 was originally designed with Windows platform in mind. ( and, by the time, when 16 bits binaries was still the norm, presumably). You had the ability to define many kinds of pointers ( e.g. far pointers vs short pointers). Hence the assumption these defines could change.
If we consider Today's platforms share, you mainly have Windows and unix _ish_ ones. Incidentally, the same defines apply to all platform, for all the defines.

As an evidence, feel free to checkout this code on Github: https://github.com/Mastercard/pkcs11-tools/blob/master/include/cryptoki/cryptoki.h , from my project (making free advertisement, I recognize :wink: ) that compile nicely on Windows(cross-compiled), Linux, Freebsd, AIX, Solaris, MacOS,... All sharing the same header file.

Actually, PKCS#11 screwed up with one thing: defining `CK_ULONG` as `unsigned int long`, which yields a 64 bits wide integer on unix _ish_ patforms. This is not an issue when the PKCS#11 drivers are all compiled through the same compiler that we use; it leads to discrepancies when e.g. you start using Cygwin64 - which implements gcc 64 bits like on a unix platform, and try to interface that code with MSVC-compiled PKCS#11 DLLs, where `sizeof(unsigned long int)` is always 32 bits, irrespective of the processor bit width.

Hope this answers your question.

Reply:
Thanks for that detailed answer.